x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -603,6 +606,106 @@ protected LogicalPlan rule(OrderBy orderBy) {
         }
     }
 
+    /**
+     * Remove unused columns created in the plan, in fields inside eval or aggregations inside stats.
+     */
+    static class PruneColumns extends Rule<LogicalPlan, LogicalPlan> {
+
+        @Override
+        public LogicalPlan apply(LogicalPlan plan) {
+            var used = new Holder<>(new AttributeSet());
+            // don't Evals without any Project/Aggregate (which might not occur as the last node in the plan)
+            var seenProjection = new Holder<>(Boolean.FALSE);
+
+            // start top-to-bottom
+            // check the differences between output and references on one side and the input on the other
+            // the extra stuff is unused and is kept
+            var pl = plan.transformDown(p -> {
+                // skip nodes that simply pass the input through
+                if (p instanceof Limit) {
+                    return p;
+                }
+
+                // remember used
+                var usedSet = used.get();
+                boolean recheck;
+                // analyze the unused items against dedicated 'producer' nodes such as Eval and Aggregate
+                // perform a loop to retry checking if the current node is completely eliminated
+                do {
+                    recheck = false;
+                    if (p instanceof Aggregate aggregate) {
+                        var pruned = seenProjection.get() ? removeUnused(aggregate.aggregates(), usedSet) : null;
+                        // no aggregates, no need
+                        if (pruned != null) {
+                            if (pruned.isEmpty()) {
+                                recheck = true;
+                                p = aggregate.child();
+                            } else {
+                                p = new Aggregate(aggregate.source(), aggregate.child(), aggregate.groupings(), pruned);
+                            }
+                        }
+
+                        seenProjection.set(Boolean.TRUE);
+                    } else if (p instanceof Eval eval) {
+                        var pruned = seenProjection.get() ? removeUnused(eval.fields(), usedSet) : null;
+                        // no fields, no eval
+                        if (pruned != null) {
+                            if (pruned.isEmpty()) {
+                                p = eval.child();
+                                recheck = true;
+                            } else {
+                                p = new Eval(eval.source(), eval.child(), pruned);
+                            }
+                        }
+                    } else if (p instanceof Project) {
+                        seenProjection.set(Boolean.TRUE);
+                    }
+                } while (recheck);
+
+                var inUse = usedSet.combine(references(p));
```

Comment:
question: since `used` only ever grows, is it possible that pruning does not correctly handle cases when a variable gets shadowed?

E.g. in the query
```
from test | eval x = salary | eval x = emp_no | keep x
```
we could prune the first eval because its value is never used.

However, it looks like this will not happen (did not check for myself) because `used` will already contain `x` once we traverse the LogicalPlan to the `eval x = salary` step.

Looking at the definition of attributes, my example might still work because the two definitions of `x` result in different attributes. However, something like
```
from test | eval x = salary | eval x = salary | keep x
```
could result in two identical attributes `x`, so that probably the first `eval x` does not get pruned, even though it is shadowed and thus unused in the output.

If I'm correct, then we'd probably need to remedy this by setting `used` to `references(p)` rather than growing `used`.

Reply:
I just realized that my suggestion at the end makes no sense because `references` does not return all required input columns; what I had in mind was some way to statically compute the required columns to execute the current `LogicalPlan` node, rather than passing down a set obtained from optimizing the downstream `LogicalPlan` node.

Reply:
There are tests with duplicated evals that test this behavior however I added an explicit one just in case.

> since used only ever grows

That's a feature not a bug - once an attribute is used, it cannot be removed, e.g.
```
 from x
| where a > 1
| eval x = b
| keep c
```

c and a need to be used, while x is removed.

Reply:
Thanks for clarifying!